### PR TITLE
E2E Testing

### PR DESCRIPTION
**Pull request by mmamayka**
The PR reworks current simulator build system to make it capable running e2e tests. 

It adds support of building the single project parts with multiple toolchains with a `ExternalProject` hack (?). See here: https://discourse.cmake.org/t/using-multiple-toolchains-in-the-same-project/1505. We need this to run e2e test that should be build with RISCV arch.

Also it adds dummy e2e test that can be run manually (it's temporary I believe)

### CODE REVIEW

**Comment by levBagryansky** [comment]
@mmamayka tests fail

**Comment by mmamayka** [comment]
> @mmamayka tests fail


Yep, thanks. I'm working on fixes

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,64 +1,101 @@
...
-add_subdirectory(third_party)
-
-enable_testing()
-include (CTest)
-if (BESM666_ENABLE_VALGRIND)
-    set(CMAKE_CTEST_ARGUMENTS --test-action memcheck)
-endif ()
-add_subdirectory(test)
+if (DEFINED BESM666__E2E_TESTS_BUILD)
+    enable_testing()
```
Comment (levBagryansky) [inline CMakeLists.txt:99]: @mmamayka This command enables all tests, it means the same as 89th line. Is it?

File: Makefile [hunk 2]
```diff
@@ -24,19 +22,32 @@ format:
 	./tools/clang-format.sh $(PWD)/$(STANDALONE_DIR)
 
 .PHONY: build
-build: init
-	cmake -S $(PWD) -B $(PWD)/$(BUILD_DIR) -DBESM666_ENABLE_VALGRIND=0
-	make -C $(PWD)/$(BUILD_DIR) -j $(JOBS)
+build:
+	cmake -S $(PWD) -B $(PWD)/$(BUILD_DIR)
```
Comment (levBagryansky) [inline Makefile:26]: @mmamayka lets add `-DBESM666_TEST_WITH_VALGRIND=OFF` to avoid building valgrind tests, for example
```
make build-val
make build
```

Comment (mmamayka) [reply Makefile:26]: They will be built only if `BESM666_TEST_WITH_VALGRIND` **defined**. So, there is not reason to set up it to `OFF`.

File: CMakeLists.txt (hunk 1 above)
Comment (mmamayka) [reply CMakeLists.txt:99]: It enables testing for e2e test built for RISCV architecture. There are 2 states in CMake building script
- build the simulator and unit tests for x86-64 triggered by defining `BESM666__SIMULATOR_BUILD`
- build e2e tests for RISCV triggered by defining `BESM666__E2E_TESTS_BUILD`
This 2 states invoked as external project consequently by the script part in `NOT BESM666__PARENT_BUILD` branch . This need to achieve capability of building different parts of project (the simulator and e2e tests in our case) with different toolchains, because CMake does not support this kind of configuration by default. So, we are using a dirty hack a bit.

So, in conclusion, we need to enable testing in different branches for different purposes =).

File: README.md [hunk 3]
```diff
@@ -1,3 +1,47 @@
...
+# Building & running E2E tests
+
+First of all, before building E2E tests you should build & install
+RISCV toolchain (its default search path is `../sysroot` relatively
+to the simulator root directory, but it can be changed by setting
+`RISCV_SYSROOT` variable before make invokation (same as `JOBS` variable)).
+
+So, the default installation script is next:
+```
+mkdir BESM-666
```
Comment (levBagryansky) [inline README.md:31]: @mmamayka What about adding dedicated script for it

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,64 +1,101 @@
 cmake_minimum_required(VERSION 3.22.1)
 
-project(BESM-666 LANGUAGES CXX)
+project(BESM-666 
+    LANGUAGES ASM CXX
+    VERSION 1.0.0
```
Comment (c71n93) [inline CMakeLists.txt:5]: Mb 0.1.0? This version don't looks like it can be released.

**Review by levBagryansky — CHANGES_REQUESTED** [review]
@mmamayka please check valgriand tests are passing

File: CMakeLists.txt (hunk 4 above)
Comment (mmamayka) [reply CMakeLists.txt:5]: This version is about to release. The next 2 PRs will fix all known errors & validate the simulator pipeline. So, let's leave it as is.

File: README.md (hunk 3 above)
Comment (mmamayka) [reply README.md:31]: Just copy and paste it, because
- I don't want to add RISCV toolchain in third party dependencies. 
- This commands sequence will be executed once during project setting up
- The root folder of this building script differs from the BESM-666 root folder

So, I suppose to avoid development of script using BESM-666 root parent directory

File: Makefile [hunk 5]
```diff
@@ -24,19 +22,32 @@ format:
...
+
+.PHONY: build-val
+build-val:
+	cmake -S $(PWD) -B $(PWD)/$(BUILD_DIR) -DBESM666_TEST_WITH_VALGRIND=ON
+	cmake --build $(PWD)/$(BUILD_DIR) --parallel $(JOBS)
+
+.PHONY: build-e2e
+build-e2e:
+	cmake -S $(PWD) -B $(PWD)/$(BUILD_DIR) -DBESM666_WITH_E2E_TESTS=ON \
+		-DBESM666_RISCV_SYSROOT=$(RISCV_SYSROOT)
```
Comment (c71n93) [inline Makefile:37]: Could you make some `if` statement on `-DBESM666_RISCV_SYSROOT`? Because in root cmake you checking this:
```
if(DEFINED BESM666_WITH_E2E_TESTS AND NOT DEFINED BESM666_RISCV_SYSROOT)
    message(FATAL_ERROR "BESM666_RISCV_SYSROOT should be defined")
endif()
```
But here you are just passing default value, that is most likely incorrect. I think we need to print error, if someone doesn't pass `RISCV_SYSROOT` value.

Comment (c71n93) [reply Makefile:37]: And this error will be printed by cmake.

**Review by levBagryansky — APPROVED** [review]
@mmamayka

**Review by c71n93 — APPROVED** [review]
